t/t0001-init.sh

```diff
@@ -464,4 +464,20 @@ test_expect_success MINGW 'redirect std handles' '
 	grep "Needed a single revision" output.txt
 '
 
+test_expect_success 'custom default branch name from config' '
+  git config --global init.defaultbranchname nmb &&
+	git init custom-config &&
+	git config --global --unset init.defaultbranchname &&
+	grep nmb custom-config/.git/HEAD
+'
+
+test_expect_success 'custom default branch name from env' '
+	GIT_DEFAULT_BRANCH_NAME=nmb git init custom-env &&
+	grep nmb custom-env/.git/HEAD
+'
```

Comment:
I believe that your logic currently has the config option overriding the environment variable. I believe that is the wrong direction, but you should make the behavior explicit in a test. (I make this suggestion because I have hit that exact problem before.)

Reply:
Yes, the common pattern in Git is that if there are environment variables _and_ config settings for the same feature, the environment variable overrides the config setting.